### ISSUE TITLE
Idea behind using top-k loss function

### ISSUE THREAD

**Issue by Ajaz-Ahmad** [mahmoodlab/CLAM#25]
Hi,

Thanks for providing such a wonderful work with Whole Slide Images. I am curious to know about few challenges and the way to tackle it.
My first question is to know your intuition behind using top-k loss function instead of Cross Entropy loss function and my second question is - I myself is solving the similar problem using feature based approach but since WSIs are too large and lesions are small, the feature space seems sparse and therefore, model highlights lots of false alarms even for negative slides. Can top-k help in better localizing the lesion in WSI.

I will really appreciate if you can clear my queries.
Thank you.

**Comment by fedshyvana** [comment]
Hi, 
Just for clarification, we used top-1 smooth svm loss only for the auxilliary objective in CLAM, which operates at a patch-level, using pseudo-labels. The pseudo labels are expected to be somewhat noisy, and it has been empirically shown that the margin-based smooth svm loss outperforms cross-entropy loss when data is limited/noisy, which is what motivated our choice. On the other hand, for training the model to perform classification at the slide-level, we do in fact use the cross-entropy loss. Please refer to our manuscript for details. 

On the other hand, I am not sure what you mean by "model highlights lots of false alarms for negative slides", are you trying to perform classification at the slide-level, patch-level, using annotations/without annotations? or are you doing object detection   or some other type of tasks?

**Comment by Ajaz-Ahmad** [comment]
Hi,

Thanks for your response. I didn't explain my problem quite well, I will try to be specific and more clear here.
I am using PANDA, only Radboud dataset-> (https://www.kaggle.com/c/prostate-cancer-grade-assessment) to predict slide labels without using annotations of individual patches. So, slide metadata (Positive or negative) is the only label I have to use. So, the data set I constructed has two labels positive (contains lesion) and negative (no lesion).
I tried with simple attention model and trained on slide level labels using feature extractor ResNet50. I am getting good accuracy on the slide level prediction. However, when it comes to highlighting the tiles with most attention I see lots of areas being highlighted.
The main regions in the slide data are stroma(60-70% in slide), healthy epithelium (30% in normal cases and decreases with severity of unhealthy cases) and unhealthy epithelium (more in unhealthy and no presence in negative slides). However, if I train ResNet50 on a supervised data set to learn two classes normal epithelium vs unhealthy epithelium. The results are good and localization (highlighting the correct regions)
I was thinking if I want to achieve better localization can Instance loss do the job?

**Comment by fedshyvana** [comment]
If I understand correctly, you're using an attention-based model, so then the attention should tell you the regions most relevant to the model's prediction and not necessarily if those regions are lesions or not. So naturally, for a negative slide, the attention will still highlight some regions with high attention because they're what the model uses for prediction - it does not mean the model is predicting them as lesion. So for tumor localization, you should only evaluate the attention-based model on positive slides. If you think localization is important in addition to slide-level classification, and you want a model that precisely scores each individual tile in the slide as positive/negative with high accuracy then a simple attention-based approach might not be what you want. You could however, try to simultaneously supervise the model with a slide-level loss and a patch-level loss, where the patch-level loss can be setup using pseudo-labels from high-attention regions or annotation.

**Comment by Ajaz-Ahmad** [comment]
Hi,

Thanks for your response. I understand your perspective. I will apply the instance loss function (the idea from your script) and see if this works. I will let you know if I've further queries.

**Comment by Ajaz-Ahmad** [comment]
Hi Again,

I have one more question. I got confused after going deeper in the script.

The problem statement I have is given slide labels [0 (negative) and 1 (positive)], I want to generate pseudo labels based on attention scores.
Now the number of classes as per the script : n_classes = 2 for camelyon data set
based on this information the Clam_SB creates two classifier branches  [line 92 in model_clam.py]. Following on this, at line 165 in the same script, the instance classifier is looped based on the number of classes. The instance label when equals 1 then instance loss is computed. However, the doubt I have is why two classifier is required when instance loss is calculated only when instance label is 1.
For example, say I have positive slide [which contains lesion and normal]
now from the slide level label I will mark each instance as positive, which is fed in first classifier out of two and then picking most attended and least attended tiles to classify.
I don't get why do we feed the same data and info into second classifier.

Am I understanding it wrong?
thanks in advance
